File: jar-common/src/test/java/com/beancounter/common/TestPositionBuckets.java
package com.beancounter.common;

import static org.assertj.core.api.Assertions.assertThat;

import com.beancounter.common.contracts.PositionRequest;
import com.beancounter.common.contracts.PositionResponse;
import com.beancounter.common.model.Currency;
import com.beancounter.common.model.MoneyValues;
import com.beancounter.common.model.Positions;
import com.beancounter.common.utils.PortfolioUtils;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.util.ArrayList;
import org.junit.jupiter.api.Test;

class TestPositionBuckets {

  @Test
  void is_DefaultMoneyValuesSet() {
    MoneyValues moneyValues = MoneyValues.builder()
        .currency(Currency.builder().code("USD").build())
        .build();

    assertThat(moneyValues).hasNoNullFieldsOrPropertiesExcept("priceData");
  }

  @Test
  void is_PositionSerializing() throws JsonProcessingException {
    ObjectMapper objectMapper = new ObjectMapper();
    PositionRequest positionRequest = PositionRequest.builder()
        .portfolioId("ABC")
        .trns(new ArrayList<>())
        .build();

    String json = objectMapper.writeValueAsString(positionRequest);
    assertThat(objectMapper.readValue(json, PositionRequest.class))
        .isEqualToComparingFieldByField(positionRequest);

    PositionResponse positionResponse = PositionResponse.builder()
        .data(new Positions(PortfolioUtils.getPortfolio("ABC")))
        .build();

    json = objectMapper.writeValueAsString(positionResponse);
    assertThat(objectMapper.readValue(json, PositionResponse.class))
        .isEqualToComparingFieldByField(positionResponse);

  }
}

File: jar-common/src/test/kotlin/com/beancounter/common/TestPositionBuckets.kt
package com.beancounter.common

import com.beancounter.common.contracts.Payload
import com.beancounter.common.contracts.PositionRequest
import com.beancounter.common.contracts.PositionResponse
import com.beancounter.common.model.Positions
import com.beancounter.common.utils.BcJson.Companion.objectMapper
import com.beancounter.common.utils.PortfolioUtils.Companion.getPortfolio
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

internal class TestPositionBuckets {
    @Test
    fun is_PositionSerializing() {
        val positionRequest = PositionRequest("ABC", ArrayList())
        var json: String = objectMapper.writeValueAsString(positionRequest)
        assertThat(objectMapper.readValue(json, PositionRequest::class.java))
            .usingRecursiveComparison().isEqualTo(positionRequest)
        val positionResponse = PositionResponse(Positions(getPortfolio()))
        json = objectMapper.writeValueAsString(positionResponse)
        assertThat(objectMapper.readValue(json, PositionResponse::class.java))
            .isNotNull
            .hasFieldOrProperty(Payload.DATA)
    }
}
